### PR TITLE
planner: retrieve optimizer hint scope change

### PR DESCRIPTION
**Pull request by foreyes**
### What problem does this PR solve? 

Retrieve change from https://github.com/pingcap/tidb/pull/11740

### What is changed and how it works?

Retrieve change from https://github.com/pingcap/tidb/pull/11740

### Check List 

Tests 

 - Unit test

Code changes

 - None

Side effects

 - None

Related changes

 - None

Release note

 - None

### CODE REVIEW

File: planner/core/logical_plan_builder.go [hunk 1]
```diff
@@ -2054,10 +2050,9 @@ func (b *PlanBuilder) TableHints() *tableHintInfo {
 func (b *PlanBuilder) buildSelect(ctx context.Context, sel *ast.SelectStmt) (p LogicalPlan, err error) {
 	b.pushSelectOffset(sel.QueryBlockOffset)
 	defer b.popSelectOffset()
-	if b.pushTableHints(sel.TableHints, typeSelect, sel.QueryBlockOffset) {
-		// table hints are only visible in the current SELECT statement.
-		defer b.popTableHints()
-	}
+	b.pushTableHints(sel.TableHints, typeSelect, sel.QueryBlockOffset)
+	// table hints are only visible in the current SELECT statement.
+	defer b.popTableHints()
```
Comment (lamxTyler) [inline planner/core/logical_plan_builder.go]: Merge two defer?

Comment (foreyes) [reply planner/core/logical_plan_builder.go]: You mean this?
```
defer func() {
    b.popSelectOffset()
    b.popTableHints()
} ()
```

Comment (lamxTyler) [reply planner/core/logical_plan_builder.go]: Yes.

File: planner/core/physical_plan_test.go [hunk 2]
```diff
@@ -660,6 +660,94 @@ func (s *testPlanSuite) TestUnmatchedTableInHint(c *C) {
...
+	_, err = se.Execute(context.Background(), "use test")
+	c.Assert(err, IsNil)
+
+	tests := []struct {
+		sql  string
+		best string
+	}{
+		// join hints
+		{
+			sql:  "select /*+ SM_JOIN(t1) */ t1.a, t1.b from t t1, (select /*+ INL_JOIN(t3) */ t2.a from t t2, t t3 where t2.a = t3.c) s where t1.a=s.a",
```
Comment (lamxTyler) [inline planner/core/physical_plan_test.go]: Could you put them into file? It will be easier for others to update results.

**Comment by foreyes** [comment]
PTAL. @lzmhhh123

File: planner/core/logical_plan_builder.go [hunk 3]
```diff
@@ -2705,11 +2702,12 @@ func buildColumns2Handle(
 
 func (b *PlanBuilder) buildUpdate(ctx context.Context, update *ast.UpdateStmt) (Plan, error) {
 	b.pushSelectOffset(0)
-	defer b.popSelectOffset()
-	if b.pushTableHints(update.TableHints, typeUpdate, 0) {
+	b.pushTableHints(update.TableHints, typeUpdate, 0)
+	defer func() {
+		defer b.popSelectOffset()
```
Comment (lzmhhh123) [inline planner/core/logical_plan_builder.go]: The `defer` in the defer function looks a little strange. Is only one defer OK?

**Comment by foreyes** [comment]
coding mistake fixed... @lzmhhh123

**Pull request merged**